dlt/destinations/sql_jobs.py

```diff
@@ -333,47 +388,59 @@ def gen_merge_sql(
                 )
             )
 
-            # create temp table used to deduplicate, only when we have primary keys
-            if primary_keys:
+        # get name of column with hard_delete hint, if specified
+        not_deleted_cond: str = None
+        hard_delete_col = get_first_column_name_with_prop(root_table, "hard_delete")
+        if hard_delete_col is not None:
+            # any value indicates a delete for non-boolean columns
+            not_deleted_cond = f"{escape_id(hard_delete_col)} IS NULL"
+            if root_table["columns"][hard_delete_col]["data_type"] == "bool":
+                # only True values indicate a delete for boolean columns
+                not_deleted_cond += f" OR {escape_id(hard_delete_col)} = {escape_lit(False)}"
+
+        # get name of column with dedup_sort hint, if specified
+        dedup_sort_col = get_first_column_name_with_prop(root_table, "dedup_sort")
+
+        insert_temp_table_name: str = None
+        if len(table_chain) > 1:
+            if len(primary_keys) > 0 or (len(primary_keys) == 0 and hard_delete_col is not None):
```

Comment:
`if len(primary_keys) > 0 or hard_delete_col is not None:` should be sufficient.